sql/catalog/tabledesc,migrations: fix namespace migration for upgraded clusters

ajwerner opened the pull request:
See individual commits.

ajwerner committed:
sql/catalog/tabledesc: update the in-memory name of system.namespace

In 21.2 we remove the existence of the deprecated namespace table which
carried the name "namespace". The existence of that old table lead to the
need for the newer, 3-part namespace table to internally carry the name
"namespace2". Internally there were hacks to make this all work. These
hacks were removed in #65340/#66064. Also, in new cluster, the namespace
table is now "namespace". However, those patches did not deal with the
old cluster and the name in their namespace table. To deal with this,
first, we add code to make it so that in-memory, even in the mixed-version
state, we can read the namespace table on new nodes.

No release note because this code has not yet seen the light of day.

Release note: None

ajwerner committed:
migrations: rewrite namespace table name in migration

In #65340 we added a migration to remove the old namespace
table which carried the name namespace2. In #66064 we removed the fact
that it existed and the hacks we had in place to deal with the name mismatch.
This posed a problem for cluster which carried the old name on the newer
namespace table. This commit migrates that name to the ideal name of
"namespace". This allows us to remove the post-deserialization changes in
the subsequent release.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66322)
<!-- Reviewable:end -->

postamar commented on pkg/sql/catalog/catconstants/namespace.go:
```diff
@@ -18,4 +18,12 @@ const (
 	NamespaceTablePrimaryIndexID = 1
+
+	// NamespaceTableName is the name of the namespace table.
+	NamespaceTableName = "namespace"
```
Perhaps reference this constant in the definition of `systemschema.NamespaceTable`?

ajwerner replied:
I do?

postamar commented on pkg/sql/catalog/tabledesc/table_desc_builder.go:
```diff
@@ -489,3 +493,15 @@ func maybeUpgradeIndexFormatVersion(idx *descpb.IndexDescriptor) (hasChanged boo
+// maybeUpgradeNamespaceName deals with upgrading the name field of the
+// namespace table (30) to be "namespace" rather than "namespace2". This
+// occurs in clusters which were bootstrapped before 21.2 and have not
+// run the corresponding migration.
```
Does this need a TODO comment to indicate the cluster version following which this can be removed?

ajwerner replied:
I figured that it would happen when the field it sets gets removed (that thing carries a TODO).

postamar reviewed (approved):
LGTM, thanks for doing this.

ajwerner commented:
TFTR!

bors r=postamar

ajwerner committed:
server: fix test which panicked due to unclean shutdown

https://teamcity.cockroachdb.com/viewLog.html?buildId=3078200&buildTypeId=Cockroach_EssentialCi&tab=buildResultsDiv

```
I210614 04:04:17.049787 152152 server/auto_upgrade.go:75  [n1] 958  error when finalizing cluster version upgrade: set-version: context canceled
panic: pebble: closed

goroutine 152180 [running]:
github.com/cockroachdb/pebble.(*DB).newIterInternal(0xc005190800, 0x0, 0x0, 0xc0009ec2d8, 0x4837790)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/db.go:708 +0x365
github.com/cockroachdb/pebble.(*DB).NewIter(0xc005190800, 0xc0009ec2d8, 0xc005b62040)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/db.go:885 +0x3d
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).init(0xc0009ec2d0, 0x4d9a740, 0xc005190800, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:170 +0x4a4
github.com/cockroachdb/cockroach/pkg/storage.newPebbleIterator(0x4d9a740, 0xc005190800, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:90 +0xd8
github.com/cockroachdb/cockroach/pkg/storage.(*Pebble).NewMVCCIterator(0xc008c4c750, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble.go:731 +0x1e5
github.com/cockroachdb/cockroach/pkg/storage.newMVCCIterator(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:830
github.com/cockroachdb/cockroach/pkg/storage.MVCCGet(0x4dc0400, 0xc00c610510, 0x4e278e0, 0xc008c4c750, 0xc0083ac116, 0x6, 0x6, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:860 +0x11a
github.com/cockroachdb/cockroach/pkg/storage.MVCCGetProto(0x4dc0400, 0xc00c610510, 0x4e278e0, 0xc008c4c750, 0xc0083ac116, 0x6, 0x6, 0x0, 0x0, 0x4e11440, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:733 +0xcc
github.com/cockroachdb/cockroach/pkg/kv/kvserver.ReadClusterVersion(0x4dc0400, 0xc00c610510, 0x4e278e0, 0xc008c4c750, 0xc008c4c750, 0xc00943a280, 0xa0, 0x98)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:2904 +0x1f7
github.com/cockroachdb/cockroach/pkg/kv/kvserver.SynthesizeClusterVersionFromEngines(0x4dc0400, 0xc00c610510, 0xc004604400, 0x1, 0x1, 0x100000015, 0xc00000000, 0x200000014, 0x0, 0xc00004e000, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:346 +0x11c
github.com/cockroachdb/cockroach/pkg/server.(*migrationServer).BumpClusterVersion(0xc00a45f290, 0x4dc0400, 0xc00c610510, 0xc0018d1130, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/migration.go:97 +0x257
github.com/cockroachdb/cockroach/pkg/server/serverpb._Migration_BumpClusterVersion_Handler.func1(0x4dde780, 0xc00d373fb0, 0x3ffce40, 0xc0018d1130, 0x6d34880, 0xc001f39760, 0x1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/serverpb/migration.pb.go:501 +0x89
github.com/cockroachdb/cockroach/pkg/util/tracing.ServerInterceptor.func1(0x4dde780, 0xc00d373fb0, 0x3ffce40, 0xc0018d1130, 0xc0099f1c40, 0xc0099f1ca0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/grpc_interceptor.go:138 +0x3d0
google.golang.org/grpc.getChainUnaryHandler.func1(0x4dc0400, 0xc00c610480, 0x3ffce40, 0xc0018d1130, 0xc001713990, 0xb70f48, 0x3f08ae0, 0xc00bb243c0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1127 +0xe7
github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1(0x4dc0400, 0xc00c610480, 0x3ffce40, 0xc0018d1130, 0xc0099f1c40, 0xc00bb243c0, 0xc00bb243c0, 0x4, 0xc00df096d0, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:173 +0xa8
google.golang.org/grpc.getChainUnaryHandler.func1(0x4dc0400, 0xc00c610480, 0x3ffce40, 0xc0018d1130, 0x0, 0xc003068400, 0x0, 0xc001f42600)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1127 +0xe7
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor(0x4dc0400, 0xc00c610480, 0x3ffce40, 0xc0018d1130, 0xc0099f1c40, 0xc00bb24380, 0xc001713ac8, 0xb70f48, 0x3f08ae0, 0xc00bb24380)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:71 +0xb9
google.golang.org/grpc.chainUnaryServerInterceptors.func1(0x4dc0400, 0xc00c610480, 0x3ffce40, 0xc0018d1130, 0xc0099f1c40, 0xc0099f1ca0, 0xc001713ba0, 0x53bc66, 0x3f213c0, 0xc00c610480)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1113 +0xd0
github.com/cockroachdb/cockroach/pkg/server/serverpb._Migration_BumpClusterVersion_Handler(0x3f19240, 0xc00a45f290, 0x4dc0400, 0xc00c610480, 0xc001f425a0, 0xc0020ba8c0, 0x4dc0400, 0xc00c610480, 0xc005284000, 0xc)
```

Release note: None

ajwerner commented:
bors r+
